### PR TITLE
[IOPID-660] : Adding Retry Fetch

### PR DESCRIPTION
**Pull request by claudiodall**
## Short description

Adding the functionality for managing multiple retries in case of timeout (defined threshold) or received specifica status code.

## List of changes proposed in this pull request

- Fetch Multiple Retries
- Adding API Fetch Param to ENV File

## How to test

Start the project, set the environment variables, for example as follow:

NEXT_PUBLIC_API_FETCH_TIMEOUT=5000
NEXT_PUBLIC_API_FETCH_MAX_RETRY=5

In Mockoon, set the delay time to 6000 for the "/profile" endpoint. 

Perform the log-in, verify that the /profile endpoint is called 5 times, as it exceeds the timeout limit.
Repeat the same operation by setting the response for the "/profile" endpoint to 500.

### CODE REVIEW

File: src/app/[locale]/_utils/api-utils.ts [hunk 1]
```diff
@@ -1,19 +1,36 @@
...
 import { ROUTES } from './routes';
+//
+// Returns a fetch wrapped with timeout and retry logic
+//
+const API_TIMEOUT = Number(`${process.env.NEXT_PUBLIC_API_FETCH_TIMEOUT}`) as Millisecond;
+const API_MAX_RETRY = Number(`${process.env.NEXT_PUBLIC_API_FETCH_MAX_RETRY}`);
 
 /** Return the implementation of fetch configured with a timeout */
-export const buildFetchApi = (
-  timeoutMs: number = 300000
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/api-utils.ts:12]: Why not leaving timeoutMs as parameter defaulting to `API_TIMEOUT`?

File: src/app/[locale]/_utils/api-utils.ts [hunk 2]
```diff
@@ -72,3 +89,43 @@ export const extractResponse = async <R>(
...
+  const abortableFetch = AbortableFetch(agent.getHttpFetch(process.env));
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const timeoutFetch: (input: any, init?: RequestInit | undefined) => Promise<Response> = toFetch(
+    setFetchTimeout(API_TIMEOUT, abortableFetch)
+  );
+  // configure retry logic with default exponential backoff
+  // @see https://github.com/pagopa/io-ts-commons/blob/master/src/backoff.ts
+  const exponentialBackoff = calculateExponentialBackoffInterval();
+  const retryLogic = withRetries<Error, Response>(API_MAX_RETRY, exponentialBackoff);
+  const retryWithError = retryLogicForResponseError((_: Response) => _.status === 500, retryLogic);
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/api-utils.ts:107]: What about passing `exponentialBackoff`, `API_MAX_RETRY` and status code to retry with as arguments defaulting to standard values (the values you are using inside the function)? What do you think about it?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: src/app/[locale]/_utils/api-utils.ts (hunk 1 above)
Comment (claudiodall) [reply src/app/[locale]/_utils/api-utils.ts:12]: Leaving now **timeoutMs** as parameter default

File: src/app/[locale]/_utils/api-utils.ts [hunk 3]
```diff
@@ -72,3 +88,58 @@ export const extractResponse = async <R>(
...
+ * @param maxRetries ( Default ENV API_MAX_RETRY ) - Maximum number of retry attempts for the request.
+ * @param statusCodeRetry ( Default 500 ) - HTTP status code that should trigger the retry.
+ * @param backOff ( Default 3000 ) - Waiting interval between retry attempts in milliseconds, this parameter represents the waiting time between each attempt with exponential growth.
+ * @returns A function that performs the fetch request with retries.
+ */
+
+export function retryingFetch(
+  maxRetries: number = API_MAX_RETRY,
+  statusCodeRetry: number = 500,
+  backOff: Millisecond = 3000 as Millisecond
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/api-utils.ts:103]:
```suggestion
  backoffBaseInterval: Millisecond = 3000 as Millisecond
```

File: src/app/[locale]/_utils/api-utils.ts [hunk 4]
```diff
@@ -72,3 +88,58 @@ export const extractResponse = async <R>(
...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const timeoutFetch: (input: any, init?: RequestInit | undefined) => Promise<Response> = toFetch(
+    setFetchTimeout(API_TIMEOUT, abortableFetch)
+  );
+  // configure retry logic with default exponential backoff
+  // @see https://github.com/pagopa/io-ts-commons/blob/master/src/backoff.ts
+  // const exponentialBackoff = calculateExponentialBackoffInterval();
+  const retryLogic = withRetries<Error, Response>(
+    maxRetries,
+    calculateExponentialBackoffInterval(backOff)
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/api-utils.ts:116]: Before of this modification we were using the default value of [500](https://github.com/pagopa/io-ts-commons/blob/a2731fd0f73faaaae269123397d0df030812eff0/src/backoff.ts#L10) ms. Am I right? Why now we changed it to `3000`?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: src/app/[locale]/_utils/api-utils.ts (hunk 4 above)
Comment (claudiodall) [reply src/app/[locale]/_utils/api-utils.ts:116]: Just, for testing, fixed, sorry

**Pull request merged**